skore/src/skore/dataset_utils.py

```diff
@@ -0,0 +1,315 @@
+    >>> y_new = [0, 1, 1]
+    >>> 
+    >>> # Check if "new" data overlaps with training data before using it
+    >>> leakage_check = compare_datasets(X_train, X_new)
+    >>> 
+    >>> # Only use for evaluation if overlap is minimal
+    >>> model = RandomForestClassifier()
+    >>> report = EstimatorReport(model, X_train=X_train, y_train=y_train)
+    >>> if leakage_check["overlap_analysis"].get("overlap_percentage", 0) < 10:
+    >>>     report.report_metrics(X_y=(X_new, y_new))
+    >>> else:
+    >>>     print("Warning: Significant overlap detected with training data")
+    """
+    # Convert numpy arrays to pandas DataFrames if needed
+    if isinstance(dataset1, np.ndarray):
+        dataset1 = pd.DataFrame(dataset1)
+    if isinstance(dataset2, np.ndarray):
+        dataset2 = pd.DataFrame(dataset2)
+    
+    # Initialize results dictionary
+    results = {
+        "shape_comparison": {
+            "dataset1_shape": dataset1.shape,
+            "dataset2_shape": dataset2.shape,
+            "same_shape": dataset1.shape == dataset2.shape
+        },
+        "column_comparison": {},
+        "overlap_analysis": {}
+    }
+    
+    # Compare columns
+    dataset1_cols = set(dataset1.columns)
+    dataset2_cols = set(dataset2.columns)
+    
+    results["column_comparison"] = {
+        "shared_columns": list(dataset1_cols.intersection(dataset2_cols)),
+        "only_in_dataset1": list(dataset1_cols - dataset2_cols),
+        "only_in_dataset2": list(dataset2_cols - dataset1_cols),
+        "same_columns": dataset1_cols == dataset2_cols
+    }
+    
+    # Check for overlapping data
+    if id_column and id_column in dataset1.columns and id_column in dataset2.columns:
+        # Using specified ID column
+        dataset1_ids = set(dataset1[id_column])
+        dataset2_ids = set(dataset2[id_column])
+        
+        overlap_ids = dataset1_ids.intersection(dataset2_ids)
+        results["overlap_analysis"] = {
+            "overlapping_ids": list(overlap_ids),
+            "overlap_count": len(overlap_ids),
+            "overlap_percentage": round(len(overlap_ids) / len(dataset1_ids) * 100, 2)
+        }
+    else:
+        # Try to check for duplicate rows
+        try:
+            shared_columns = results["column_comparison"]["shared_columns"]
+            if shared_columns:
+                # Check overlapping rows based on shared columns
+                dataset1_subset = dataset1[shared_columns].drop_duplicates()
+                dataset2_subset = dataset2[shared_columns].drop_duplicates()
+                
+                merged = pd.merge(
+                    dataset1_subset, dataset2_subset, 
+                    on=shared_columns, how='inner'
+                )
+                
+                overlap_count = len(merged)
+                results["overlap_analysis"] = {
+                    "overlap_row_count": overlap_count,
+                    "overlap_percentage": round(overlap_count / len(dataset1) * 100, 2)
```

Comment:
Whenever it's not super plain to do, can you split the big function in sub functions please? This way, it's much easier to test all the use-cases, and also to read.